fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java

```diff
@@ -0,0 +1,194 @@
+            return false;
+        }
+
+        // check distribution satisfy scan node
+        List<Integer> groupBy = aggregate.getGroupingKeys().stream().map(ColumnRefOperator::getId)
+                .collect(Collectors.toList());
+
+        if (groupBy.isEmpty() || groupBy.containsAll(scan.getDistributionSpec().getShuffleColumns())) {
+            return false;
+        }
+
+        groupBy.add(distinctColumns.get(0).getId());
+        return groupBy.containsAll(scan.getDistributionSpec().getShuffleColumns());
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) input.getOp();
+
+        Map<ColumnRefOperator, CallOperator> distinctMap = Maps.newHashMap();
+        Map<ColumnRefOperator, CallOperator> otherMap = Maps.newHashMap();
+
+        aggregate.getAggregations().forEach((k, v) -> {
+            if (isDistinct(v)) {
+                distinctMap.put(k, v);
+            } else {
+                otherMap.put(k, v);
+            }
+        });
+
+        LogicalAggregationOperator.Builder second = new LogicalAggregationOperator.Builder();
+
+        Optional<ColumnRefOperator> distinctColumn = distinctMap.values().stream().map(f -> f.getChild(0))
+                .map(c -> (ColumnRefOperator) c).distinct().findFirst();
+
+        List<ColumnRefOperator> firstGroupBy = Lists.newArrayList(aggregate.getGroupingKeys());
+        Preconditions.checkState(distinctColumn.isPresent());
+        firstGroupBy.add(distinctColumn.get());
+
+        Map<ColumnRefOperator, CallOperator> firstAggregations = Maps.newHashMap();
+        Map<ColumnRefOperator, CallOperator> secondAggregations = Maps.newHashMap();
+
+        ColumnRefFactory factory = context.getColumnRefFactory();
+        otherMap.forEach((k, v) -> {
+            Function origin = v.getFunction();
+            String firstFn = OTHER_FUNCTION_TRANS.get(origin.getFunctionName().getFunction()).first;
+            String secondFn = OTHER_FUNCTION_TRANS.get(origin.getFunctionName().getFunction()).second;
+
+            CallOperator firstAgg = getAggregation(firstFn, origin.getArgs(), v.getChildren());
+            ColumnRefOperator firstOutput = factory.create(firstAgg, firstAgg.getType(), firstAgg.isNullable());
+
+            firstAggregations.put(firstOutput, firstAgg);
+
+            CallOperator secondAgg =
+                    getAggregation(secondFn, new Type[] {firstAgg.getType()}, Lists.newArrayList(firstOutput));
+            secondAggregations.put(k, secondAgg);
+        });
+
+        distinctMap.forEach((k, v) -> {
+            Function origin = v.getFunction();
+            String secondFn = DISTINCT_FUNCTION_TRANS.get(origin.getFunctionName().getFunction());
+            CallOperator secondAgg = getAggregation(secondFn, origin.getArgs(), v.getChildren());
+            secondAggregations.put(k, secondAgg);
+        });
```

Comment:
Distinct only has second aggregation?

Reply:
For count distinct column, first stage: group by(distinct), second stage: count the column... you can see I put the distinct column to GroupBy in first stage